src/apps/socket/dev.lua

```diff
@@ -1,38 +1,58 @@
-module(...,package.seeall)
+module(..., package.seeall)
 
+local S = require("syscall")
+local h = require("syscall.helpers")
+local packet = require("core.packet")
+local bit = require("bit")
 local ffi = require("ffi")
 local C = ffi.C
 
-local packet = require("core.packet")
-require("lib.raw.raw_h")
-require("apps.socket.io_h")
-
-dev = {}
-
-function dev:new (ifname)
-   assert(ifname)
-   self.__index = self
-   local dev = {fd = C.open_raw(ifname)}
-   return setmetatable(dev, self)
+local c, t = S.c, S.types.t
+
+RawSocketDev = {}
+
+function RawSocketDev:new (name)
+   local tp = h.htons(c.ETH_P["ALL"])
+   local sock, err = S.socket(c.AF.PACKET, bit.bor(c.SOCK.RAW, c.SOCK.NONBLOCK), tp)
+   if not sock then return nil, err end
+   local index, err = S.util.if_nametoindex(name, sock)
+   if err then
+      sock:close()
+      return nil, err
+   end
+   local addr = t.sockaddr_ll{sll_family = c.AF.PACKET, sll_ifindex = index, sll_protocol = tp}
+   local ok, err = S.bind(sock, addr)
+   if not ok then
+      S.close(sock)
+      return nil, err
+   end
+   return setmetatable({sock = sock}, {__index = RawSocketDev})
 end
 
-function dev:transmit (p)
-   assert(self.fd)
-   assert(C.send_packet(self.fd, p) ~= -1)
+function RawSocketDev:transmit (p)
+   local _, err = S.write(self.sock, p.data, p.length)
```

Comment:
> What happens if write fails to transmit p.length data?

Good question!

Breaking it down:

1. How much data should be buffered?
1. Should this be buffered by the kernel, by Snabb, or both?
1. What should happen when the buffer is full?

I would like to make a case in favor of the behavior on the PR now:

1. Buffer should be sufficient for one "burst" of packets. Ballpark 256KB?
1. The kernel already has a buffer, and we should be able to adjust its size by `setsockopt()`, so don't need to buffer in Snabb too.
1. If the buffer is full (`EAGAIN` or `EWOULDBLOCK`) then we should drop the packet immediately, provided that we trust the buffer to be big enough to handle bursts and only fill up during actual congestion (insufficient throughput from the kernel).